### PR TITLE
Adding Non-Max Suppression Layer

### LINKED ISSUE #461: NonMaxSupression Layer
This should return a KerasCV bounding box format instead of a custom object.

https://github.com/LukeWood/reef-net/blob/master/reef_net/model.py#L247

### PR DESCRIPTION
**Pull request by quantumalaviya**
# What does this PR do?
This PR will add an NMS layer that supports the KerasCV bounding box API. The layer is inspired by [Luke Wood's RetinaNet implementation](https://github.com/LukeWood/reef-net/blob/5f1a3fb8b963fb024a3c49da7ffd43bdccb636cd/reef_net/layers/decode_predictions.py#L7). The biggest change is the support for dynamically computed batch size and removal of the anchor box calculations from this layer (this should be done outside the NMS layer).

Fixes #461 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case. (#461)
- [x] Did you write any new necessary tests?

## Who can review?
@LukeWood

### CODE REVIEW

**Review by LukeWood — CHANGES_REQUESTED** [review]
Thank you for the great PR.  I've left you some minor comments.

File: keras_cv/bounding_box/non_max_suppression.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    ):
+        super().__init__(**kwargs)
+        self.num_classes = num_classes
+        self.bounding_box_format = bounding_box_format
+        self.confidence_threshold = confidence_threshold
+        self.nms_iou_threshold = nms_iou_threshold
+        self.max_detections = max_detections
+        self.max_detections_per_class = max_detections_per_class
+
+    def call(self, images, predictions):
```
[the remark this answers is not in the document]
Comment (quantumalaviya) [reply keras_cv/bounding_box/non_max_suppression.py:111]: Yeah, doesn't have to be relative for the tf nms op (both yxyx and rel_yxyx are fine).

We can make images optional but they're required for when the user passes relative formats.

File: keras_cv/bounding_box/non_max_suppression.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    1) picking the highest ranked boxes
+    2) pruning away all boxes that have a high IoU with the chosen boxes.
+
+    References:
+        - [Yolo paper](https://arxiv.org/pdf/1506.02640)
+
+    Args:
+        num_classes: an integer representing the number of classes that a bounding
+            box can belong to.
+        bounding_box_format: a case-sensitive string which is one of the supported
```
[the remark this answers is not in the document]
Comment (quantumalaviya) [reply keras_cv/bounding_box/non_max_suppression.py:34]: I agree. Can you link the docs here? I can't find the documentation for such a list.

[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/bounding_box/non_max_suppression.py:34]: https://keras.io/api/keras_cv/bounding_box/formats/

File: keras_cv/bounding_box/non_max_suppression.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/bounding_box/non_max_suppression.py:111]: Oh good point, we can actually update the converters to support no-images rel->rel conversions.

File: keras_cv/bounding_box/non_max_suppression.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                    tf.expand_dims(
+                        nmsed_boxes.nmsed_scores[i][:num_detections], axis=-1
+                    ),
+                ],
+                axis=-1,
+            )
+
+            # iterate through the boxes and append it to TensorArray
+            for i in range(nmsed_boxes.valid_detections[i]):
+                boxes = boxes.write(boxes.size(), boxes_recombined[i])
```
Comment (LukeWood) [inline keras_cv/bounding_box/non_max_suppression.py:151]: you're shadowing i here.

for i in range(...[i]).

You need to use another variable, like j.

Then I think you can replace boxes.size() with j.

File: keras_cv/bounding_box/non_max_suppression.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        # converting all boxes to the original format
+        boxes = convert_format(
+            tf.expand_dims(boxes, axis=0),
+            source="yxyx",
+            target=self.bounding_box_format,
+            images=images,
+        )[0]
+
+        # using cumulative sum to calculate row_limits for ragged tensor
+        row_limits = tf.cumsum(nmsed_boxes.valid_detections)
```
Comment (LukeWood) [inline keras_cv/bounding_box/non_max_suppression.py:165]: clever use of this

**Review by LukeWood — CHANGES_REQUESTED** [review]
Biggest concern is the variable shadowing

File: keras_cv/bounding_box/non_max_suppression.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+            scores,
+            self.max_detections_per_class,
+            self.max_detections,
+            self.iou_threshold,
+            self.confidence_threshold,
+            clip_boxes=False,
+        )
+
+        # output will be a ragged tensor because num_boxes will change across the batch
+        return self._encode_to_ragged(nmsed_boxes, images)
```
[the remark this answers is not in the document]
Comment (quantumalaviya) [reply keras_cv/bounding_box/non_max_suppression.py:126]: Yes, we can cover this as part of a follow-up PR.

File: keras_cv/bounding_box/non_max_suppression.py (hunk 3 above)
Comment (quantumalaviya) [reply keras_cv/bounding_box/non_max_suppression.py]: That's a great catch, thanks!

That being said, I have left the `boxes.size()` as it is to mimic the `append` function whereas `j` will reset for every batch.

File: keras_cv/bounding_box/non_max_suppression.py (hunk 4 above)
Comment (quantumalaviya) [reply keras_cv/bounding_box/non_max_suppression.py:165]: Thanks!

File: keras_cv/bounding_box/non_max_suppression.py [hunk 6]
```diff
@@ -147,8 +147,8 @@ def _encode_to_ragged(self, nmsed_boxes, images):
             )
 
             # iterate through the boxes and append it to TensorArray
-            for i in range(nmsed_boxes.valid_detections[i]):
-                boxes = boxes.write(boxes.size(), boxes_recombined[i])
+            for j in range(nmsed_boxes.valid_detections[i]):
+                boxes = boxes.write(boxes.size(), boxes_recombined[j])
```
Comment (LukeWood) [inline keras_cv/bounding_box/non_max_suppression.py:151]: any reason to keep using boxes.size() here now?

Comment (quantumalaviya) [reply keras_cv/bounding_box/non_max_suppression.py:151]: This is because j will be reset for every batch, we need an append function

Comment (LukeWood) [reply keras_cv/bounding_box/non_max_suppression.py:151]: Ohhh I see, I'm actually surprised that this works in graph mode.  What a wonderful trick, thanks for showing me this.

**Review by LukeWood — APPROVED** [review]

**Comment by LukeWood** [comment]
Thank you Sarvagya for the wonderful contribution.

@quantumalaviya

**Pull request merged**